### PR TITLE
Import foreign schema

### PR DESCRIPTION
**Pull request by yajirobee**
Implemented import foreign schema.
Treasure Data originally doesn't have PostgreSQL like schema. I implemented import foreign schema to import tables on the specified TD database into the specified PostgreSQL schema. I suppose this is a reasonable implementation since tables can be accessed by the syntax "\<database\>.\<table\>" in TD and "\<schema\>.\<table\>" in PostgreSQL. i.e. database of TD and schema of PostgreSQL work the same manner in the query syntax.
Please note "time" column will NOT be imported by IMPORT FOREIGN SCHEMA because TD table API does not include time column in the table schema.

### CODE REVIEW

File: bridge_td_client_rust/src/lib.rs [hunk 1]
```diff
@@ -779,3 +781,141 @@ pub extern fn import_commit(
...
+            return Err(String::from("td API error"))
+        },
+        Err(err) => {
+            log!(error_log, "import_schema: Failed to import schema. {:?}", err);
+            return Err(String::from("td API error"))
+        }
+    };
+
+    let mut commands = Vec::new();
+    'outer: for tab in &tables {
```
Comment (komamitsu) [inline bridge_td_client_rust/src/lib.rs]: [minor] This kind complicated labeled loop might be replaced with simple recursive function calls.

File: bridge_td_client_rust/src/lib.rs [hunk 2]
```diff
@@ -779,3 +781,141 @@ pub extern fn import_commit(
...
+        // build a query of create foreign table
+        let mut command = format!("CREATE FOREIGN TABLE {} (\n", tab.name);
```
Comment (komamitsu) [inline bridge_td_client_rust/src/lib.rs]: I think this library shouldn't create SQL to make the responsibility clear. How about just returning array of columns to `treasuredata_fdw.c`?

File: treasuredata_fdw.c [hunk 3]
```diff
@@ -2259,275 +2260,65 @@ analyze_row_processor(PGresult *res, int row, TdFdwAnalyzeState *astate)
...
+	ereport(DEBUG1,
+			(errmsg("apikey: %s, endpoint: %s, query_engine: %s",
+					apikey, endpoint, query_engine)));
+	ereport(DEBUG1,
+			(errmsg("remote_database: %s, server: %s",
+					stmt->remote_schema, stmt->server_name)));
+
+	/* Note that imported tables will NOT include time column
+	   since TD API does not return time column.*/
+	commands = importSchema(apikey, endpoint, query_engine, stmt->remote_schema,
```
Comment (komamitsu) [inline treasuredata_fdw.c]: SQL statement should be constructed in this file not in the Rust library.

File: bridge_td_client_rust/src/lib.rs [hunk 4]
```diff
@@ -779,3 +781,141 @@ pub extern fn import_commit(
...
+                "float" => command.push_str("real"),
+                "double" => command.push_str("double precision"),
+                "string" => command.push_str("text"),
+                s if s.starts_with("array") => {
+                    let type_elems: Vec<&str> = s.split("<").collect();
+                    let dimension = type_elems.len() - 1;
+                    let td_arr_type = type_elems[type_elems.len() - 1]
+                        .trim_right_matches(">");
+                    let pg_arr_type = match td_arr_type {
+                        "int" => "int",
```
Comment (komamitsu) [inline bridge_td_client_rust/src/lib.rs]: [minor] These lines are duplicated with above ones.

File: .travis.yml [hunk 5]
```diff
@@ -24,5 +24,5 @@ before_install: curl -sSf https://static.rust-lang.org/rustup.sh | sudo sh
 install: make && sudo make install
 
 script:
-  - 'if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then ./setup_regress presto && travis_wait make installcheck; fi'
-  - 'if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then ./setup_regress hive && travis_wait make installcheck; fi'
+  - 'if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then ./setup_regress presto && travis_wait make installcheck; if [ $? -ne 0 ]; then cat regression.diffs && false; fi; fi'
```
[the remark this answers is not in the document]
Comment (yajirobee) [reply .travis.yml]: I understand your intention. reverted.

**Comment by yajirobee** [comment]
I fixed to build import foreign table queries on treasuredata_fdw.c instead of Rust code. Can you please review again?

File: bridge_td_client_rust/src/lib.rs [hunk 6]
```diff
@@ -779,3 +797,132 @@ pub extern fn import_commit(
...
+            "float" => Ok(String::from("real")),
+            "double" => Ok(String::from("double precision")),
+            "string" => Ok(String::from("text")),
+            other_type => Err(String::from(
+                format!("Unexpected TypeName: {:?}", other_type)))
+        }
+    };
+
+    fn convert_array_type_td_to_pg(td_type_name: &str) -> Result<String, String> {
+        let type_elems: Vec<&str> = td_type_name.split("<").collect();
```
[the remark this answers is not in the document]
Comment (yajirobee) [reply bridge_td_client_rust/src/lib.rs]: Thank you for your suggestion. Obviously, your implementation is robust for unexpected cases. I use it.

**Comment by yajirobee** [comment]
Thank you for your comments! I fixed some codes. Please review them when you get a chance.

**Pull request merged**